mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +348,209 @@ propagateUnmergeVectorization(T &&dimAndLength,
   return result;
 }
 
+// Key data structures for vectorization analysis
+struct DimInfo {
+  TransformMapAttr transformMap; // Map to which the dimension belongs
+  TransformAttr transform;       // transform to which the dimension belongs
+  size_t positionInMerge; // position of the dimension inside the transform
+
+  // Utility function to return a transform pair <map,transform>
+  std::pair<TransformMapAttr, TransformAttr> transformPair() {
+    return std::make_pair(transformMap, transform);
+  }
+};
+using DimToMergeMap = llvm::SmallDenseMap<uint32_t, DimInfo>;
+
+using ContiguousMergesMap =
+    llvm::DenseMap<std::pair<TransformMapAttr, TransformAttr>,
+                   llvm::EquivalenceClasses<uint32_t>>;
+
+void findCountiguousGroupsUnmerge(const ArrayRef<uint32_t> upperDims,
+                                  const ArrayRef<int64_t> params,
+                                  DimToMergeMap &dimToMerge,
+                                  ContiguousMergesMap &contiguousGroups) {
+
+  size_t i = 0;
+  while (i < upperDims.size()) {
+    // The i-th dimension (i.e., dimension upperDims[i])
+    // uses params[i]
+
+    auto keyI = dimToMerge[upperDims[i]];
+    if (!keyI.transform) {
+      i++;
+      continue;
+    }
+
+    SmallVector<uint32_t> groupCandidate;
+    SmallVector<size_t> dimPosition;
+    for (size_t j = i; j < upperDims.size(); j++) {
+
+      // Unit lengths don't affect the mergeability logic
+      if (params[j] == 1) {
```

Comment:
I'm still not quite happy about this one, unless it's not an issue in practice. Consider
`Merge{3, 1, 8}` , aka `(d0) -> (d0 / 8, 0, d0 % 8)`.

If we call those results `(a, b, c)`, then I'd expect both
`Unmerge{3, 8}(a, c)` and `Unmerge{3, 1, 8}(a, b, c)` to lead to a finding of continuity.

Reply:
But also if we let `d` be a length-1 dimension, we might also want `Unmerge{3, 1, 1, 8}(a, b, d, c)` to work ... or maybe we don't because that's a silly pathological function no one should've been asking for.

Reply:
I have the suspicious that those cases would work (although pathological they are):
- `Merge{3, 1, 8} at [a,b,c]`  and `Unmerge{3, 1, 8}(a, b, c)` would work for the code
- `Merge{3, 1, 8} at [a,b,c]`  and `Unmerge{3, 8}(a, b, c)` would work, because I don't check for the exact position. I only check that the paramters are the same and that the position are sorted in the same way of the merge. So it should work. 
-  `Merge{3, 1, 8} at [a,b,c]`  and `Unmerge{3, 1,1,8}(a, b, c)` should.... work? I think it is the same logic as before (parameters the same and position sorted in the same way)
Should I add a test and verify?

Reply:
Ok, yes, the case `Merge{3, 1, 8}` -> `Unmerge{3, 8}(a, b, c)` would not work. 

But `Unmerge{3, 1, 1, 8}(a, b, d, c)` would, right? I am skipping over the singleton so for me that one is equivalent to `Unmerge{3,8}(a,c)` which would be ok. 

For `Merge{3, 1, 8}` -> `Unmerge{3, 8}(a, b, c)` I am not sure exactly how to make it work out elegantly. Any suggestions :) ?

Reply:
Ah, an idea could be that I don't track the dimensions in the merge that are broadcasts. I.e., when I meet `Merge{3,1,8} at [0,1,2]` I won't store `1` in my `dimToMerge` table. What do you think?

Reply:
(But hey, if it all works, it all works)

Reply:
(Some of your comments disappeared, but I'd add those three examples as tests)

Reply:
(sorry about deleted comments, I was basically thinking out loud )
Yes, you are right. If the `Merge` has a broadcast dimension we need to add it into the group. I did that, and I also  introduced a `isBroadcast` field in the dim table to track when different transformations introduce the broadcast.